sqlmigrations: set system.jobs TTL to 10m

madelynnblue opened the pull request:
Some jobs have multiple updates per minute, take hours to run,
and have a large payload because they are coordination distributed
progress among nodes. These issues taken together can result in a
single row taking up a large amount of space due MVCC. Shorten the
TTL of the system.jobs table so the old updates are cleaned up. This
should prevent the jobs table from splitting due to progress updates
on a single row, and should prevent a single row from ever getting
excessively large.

Do some refactoring to make some of this reusable.

Release note (performance): prevent the jobs table from growing too
large while updating.

See #21561

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21575)
<!-- Reviewable:end -->

benesch commented:
The test failures here suck. There's no reason it needs to be this bad. I opened https://github.com/cockroachdb/cockroach/pull/21579 to make nearly all of these tests not depend on the particular migrations that get run at bootstrap.

benesch commented:
Woohoo, looking good!

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sqlmigrations/migrations.go, line 605 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3DnvBVx05gsvMNLam5:-L3DnvBVx05gsvMNLam6:b8d4q4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations.go#L605)):*
> ```Go
> 
> 	// Retrieve the existing .meta zone config.
> 	metaZone, err := getZoneConfigByID(ctx, r, keys.MetaRangesID)
> ```

I would replace this with `getZoneConfig(ctx, r, "RANGE meta")`.

---

*[pkg/sqlmigrations/migrations.go, line 619 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3Dnz2SUODRTHjVkdzq:-L3Dnz2TkQ3RPwHNgoAO:bxjp1c7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations.go#L619)):*
> ```Go
> 	// The liveness range was previously covered by the ".system" zone. Grab the
> 	// existing ".system" zone (if any) for modification.
> 	livenessZone, err := getZoneConfigByID(ctx, r, keys.SystemRangesID)
> ```

And likewise here. `RANGE system`.

---

*[pkg/sqlmigrations/migrations.go, line 670 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3Do-1xK1FzHpmdQgKH:-L3Do-1xK1FzHpmdQgKI:b-h5cwfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations.go#L670)):*
> ```Go
> }
> 
> func getZoneConfigFull(ctx context.Context, r runner, stmtFor string) (config.ZoneConfig, error) {
> ```

Then you can ditch the two helpers and rename this thing `getZoneConfig`.

---

*[pkg/sqlmigrations/migrations.go, line 954 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3Do8eVLse40gB9IWxd:-L3Do8eVLse40gB9IWxe:blo3joe) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations.go#L954)):*
> ```Go
> 	defaultTTLSeconds := config.DefaultZoneConfig().GC.TTLSeconds
> 
> 	jobsZone, err := getZoneConfigByTable(ctx, r, "system.jobs")
> ```

(Provided you say `TABLE system.jobs` here.)

---

*[pkg/sqlmigrations/migrations_test.go, line 686 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3DoCGwqI08L4aWyHYM:-L3DoCGx5Z03-5tN84T6:b-hw90qc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations_test.go#L686)):*
> ```Go
> 	}
> 	if !reflect.DeepEqual(expected, zone) {
> 		t.Fatalf("unexpected .meta zone config: %s", pretty.Diff(expected, zone))
> ```

`.meta` is hardcoded here.

---

*[pkg/sqlmigrations/migrations_test.go, line 801 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3DoT2At40k_q5sU2Q1:-L3DoT2At40k_q5sU2Q2:bxe6qlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations_test.go#L801)):*
> ```Go
> 	expected := config.DefaultZoneConfig()
> 	expected.GC.TTLSeconds = 10 * 60
> 	checkZoneConfig(t, mt.sqlDB, keys.JobsTableID, expected)
> ```

We should add a few more assertions here, like we do for .meta and .liveness. Namely, that we use the `.system` config as a basis, and that we update an existing jobs zone config provided the TTL hasn't been changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-:-L3DoddFmtUvzF7krI1T:brxcaal)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sqlmigrations/migrations.go, line 605 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3DnvBVx05gsvMNLam5:-L3F7-byKNRSOZn9f7n1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations.go#L605)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I would replace this with `getZoneConfig(ctx, r, "RANGE meta")`.
</blockquote></details>

Done.

---

*[pkg/sqlmigrations/migrations.go, line 619 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3Dnz2SUODRTHjVkdzq:-L3F7-OsTHhw1zqpMlKl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations.go#L619)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

And likewise here. `RANGE system`.
</blockquote></details>

Done.

---

*[pkg/sqlmigrations/migrations.go, line 670 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3Do-1xK1FzHpmdQgKH:-L3F70ffnkvPneAXTQIe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations.go#L670)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Then you can ditch the two helpers and rename this thing `getZoneConfig`.
</blockquote></details>

Done.

---

*[pkg/sqlmigrations/migrations.go, line 954 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3Do8eVLse40gB9IWxd:-L3F715yLCZRfBKXZysg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations.go#L954)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

(Provided you say `TABLE system.jobs` here.)
</blockquote></details>

Done.

---

*[pkg/sqlmigrations/migrations_test.go, line 686 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3DoCGwqI08L4aWyHYM:-L3F788j_8iDygJkiu1R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations_test.go#L686)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

`.meta` is hardcoded here.
</blockquote></details>

Done.

---

*[pkg/sqlmigrations/migrations_test.go, line 801 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-L3DoT2At40k_q5sU2Q1:-L3FF-DUQofOV8a1KPJ1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5217235553c4f8e48ecde5276501ac11ba2b21/pkg/sqlmigrations/migrations_test.go#L801)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

We should add a few more assertions here, like we do for .meta and .liveness. Namely, that we use the `.system` config as a basis, and that we update an existing jobs zone config provided the TTL hasn't been changed.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575)*
<!-- Sent from Reviewable.io -->

benesch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

P.S. scripts/diff-example.sh will make your life much better while fixing these last two tests:

```
make test PKG=./pkg/cli TESTS= | scripts/diff-example.sh
make test PKG=./pkg/ccl/cliccl TESTS=zone | scripts/diff-example.sh
```

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-:-L3Ff8K7Q3gg7duZvLzU:b-hw23hb)*
<!-- Sent from Reviewable.io -->

benesch commented:
Whoops.

```
make test PKG=./pkg/cli TESTS=zone | scripts/diff-example.sh
make test PKG=./pkg/ccl/cliccl TESTS=zone | scripts/diff-example.sh
```

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21575#-:-L3FfP7uyFGvojY-FVpC:b-5fo5c8)*
<!-- Sent from Reviewable.io -->

mjibson commented:
This stuff is pretty fun. Maybe it works now.

benesch commented:
Woo, now it’s just the linter that’s unhappy!

madelynnblue committed:
sqlmigrations: set system.jobs TTL to 10m

Some jobs have multiple updates per minute, take hours to run,
and have a large payload because they are coordination distributed
progress among nodes. These issues taken together can result in a
single row taking up a large amount of space due MVCC. Shorten the
TTL of the system.jobs table so the old updates are cleaned up. This
should prevent the jobs table from splitting due to progress updates
on a single row, and should prevent a single row from ever getting
excessively large.

Do some refactoring to make some of this reusable.

Release note (performance): prevent the jobs table from growing too
large while updating.

See #21561
